Blocks: Introduce `WP_Block_Processor` for efficiently parsing blocks.

dmsnell opened the pull request:
Replaces #6760 

Trac ticket: Core-61401
See: (~#9105~), ~#9841~, ~#9842~ -> #10291, ~#9843~ -> #10292

## Todo

 - [x] Update `parse_blocks()` docs and explain when to use which interface

## Description

The Block Scanner follows the HTML API in providing a streaming,
near-zero-overhead, lazy, re-entrant parser for traversing block
structure. This class provides an alternate interface to
`parse_blocks()` which is more amenable to a number of common
server-side operations on posts, such as:

 - Generating an excerpt from only the first N blocks in a post.
 - Determining which block types are present in a post.
 - Determining which posts contain a block of a given type.
 - Generating block supports content for a post.

## Planned refactors

 - `traverse_and_serialize_blocks()` / block hooks
 - finding post galleries

Soean commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+			 * until the comment closing must be completely whitespace.
+			 */
+			if ( ! $has_json ) {
+				$max_whitespace_length = $comment_closing_at - $json_at - $void_flag_length;
```
The value of the variable is not used anywhere.

dmsnell replied:
gone!

tomjn commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 * @param string $source_text Input document potentially containing block content.
+	 * @return ?self Created block scanner, if successfully created.
+	 */
+	public static function create( $source_text ) {
```
Any reason not to type hint this with `string $source_text`? If someone isn't checking they could not realise it's returned a `null` value

dmsnell replied:
this is gone now, but I ran into issues with the static creator method, which I think we discussed in another comment. marking as resolved.

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 * @param string $source_text Input document potentially containing block content.
+	 * @return ?self Created block scanner, if successfully created.
+	 */
+	public static function create( $source_text ) {
```
```suggestion
	public static function create( $source_text ): ?self {
```

westonruter replied:
Although, if the `string` type hint is added to `$source_text` then a fatal error would have occurred, and this method would never return anything but `self`:

```suggestion
	public static function create( string $source_text ): self {
```

dmsnell replied:
leaving this open with `?self` so that we can fail it in the future if we decide to. otherwise this and every subclass will be prevented from adding conditions to creating the scanner.

dmsnell replied:
this method is gone, so this comment is no longer relevant.

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 *                                not surrounded by block delimiters. Defaults to `skip-freeform`.
+	 * @return bool Whether a block delimiter was matched.
+	 */
+	public function next_delimiter( $freeform_blocks = 'skip-freeform' ) {
```
```suggestion
	public function next_delimiter( string $freeform_blocks = 'skip-freeform' ): bool {
```

dmsnell replied:
this method has been changed and I added type annotations

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 *
+	 * @since {WP_VERSION}
+	 */
+	private function __construct( $source_text ) {
```
```suggestion
	private function __construct( string $source_text ) {
```

dmsnell replied:
added!

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 * @param int $search_end          Last offset in which to search, for limiting search span.
+	 * @return int Offset after the current HTML comment ends, or `$end` if no end was found.
+	 */
+	private function find_html_comment_end( $comment_starting_at, $search_end ) {
```
```suggestion
	private function find_html_comment_end( int $comment_starting_at, int $search_end ): int {
```

dmsnell replied:
added!

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 * @return string|null Error from last attempt at parsing next block delimiter,
+	 *                     or `NULL` if last attempt succeeded.
+	 */
+	public function get_last_error() {
```
```suggestion
	public function get_last_error(): ?string {
```

dmsnell replied:
added!

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 *
+	 * @return int JSON_ERROR_ code from last attempt to parse block JSON attributes.
+	 */
+	public function get_last_json_error() {
```
```suggestion
	public function get_last_json_error(): int {
```

dmsnell replied:
added!

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 *
+	 * @return string|null type of the block comment delimiter, if currently matched.
+	 */
+	public function get_delimiter_type() {
```
```suggestion
	public function get_delimiter_type(): ?string {
```

dmsnell replied:
added!

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 *
+	 * @return bool Whether the currently-matched block delimiter contains the closing flag.
+	 */
+	public function has_closing_flag() {
```
```suggestion
	public function has_closing_flag(): bool {
```

dmsnell replied:
added!

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 *                           E.g. "paragraph", "core/paragraph", "math-blocks/formula".
+	 * @return bool Whether this delimiter represents a block of the given type.
+	 */
+	public function is_block_type( $block_type ) {
```
```suggestion
	public function is_block_type( $block_type ): bool {
```

dmsnell replied:
added!

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 * @return bool Whether the matched block delimiter opens a block, and whether it
+	 *              opens a block of one of the given block types, if provided.
+	 */
+	public function opens_block( ...$block_type ) {
```
```suggestion
	public function opens_block( ...$block_type ): bool {
```

dmsnell replied:
added!

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 *
+	 * @return bool Whether or not the matched delimiter is implied as `core/freeform`.
+	 */
+	public function is_freeform() {
```
```suggestion
	public function is_freeform(): bool {
```

dmsnell replied:
this method is gone, but the renamed one has the type annotation.

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 * @return bool Whether the currently-matched delimiter is implicit and surround
+	 *              top-level non-block content containing non-whitespace text.
+	 */
+	public function is_non_whitespace_freeform() {
```
```suggestion
	public function is_non_whitespace_freeform(): bool {
```

dmsnell replied:
this method is gone, but the renamed form has the type annotation.

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 *
+	 * @return string Fully-qualified block namespace and type, e.g. "core/paragraph".
+	 */
+	public function get_block_type() {
```
```suggestion
	public function get_block_type(): string {
```

dmsnell replied:
added! but it can also be `null`

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 *
+	 * @return array|null Parsed JSON attributes, if present and valid, otherwise `null`.
+	 */
+	public function allocate_and_return_parsed_attributes() {
```
```suggestion
	public function allocate_and_return_parsed_attributes(): ?array {
```

dmsnell replied:
added!

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1096 @@
+	 *
+	 * @return WP_HTML_Span|null Span of text in source text spanning matched delimiter.
+	 */
+	public function get_span() {
```
```suggestion
	public function get_span(): ?WP_HTML_Span {
```

dmsnell replied:
added!

tomjn commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1244 @@
+	 *
+	 * @since 6.9.0
+	 */
+	protected function __construct( string $source_text ) {
```
Is there any reason this is protected? I see there is a create static function but the create static function appears to be doing the job this constructor is meant to do, albeit without the type safety and protections a constructor brings.

Could the doc block explain this or reference the correct way to construct a new scanner object? Most objects in WordPress use the `new` operator for object creation so this is a bit odd

dmsnell replied:
this is no longe protected

tomjn commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1244 @@
+	 * @param string $source_text Input document potentially containing block content.
+	 * @return ?self Created block scanner, if successfully created.
+	 */
+	public static function create( string $source_text ) {
```
From my comment on the commit https://github.com/WordPress/wordpress-develop/commit/85452871c61321a5c48f688fb3a94b16e051ca45#commitcomment-167116766

```suggestion
	public static function create( string $source_text ): WP_Block_Scanner {
```

Any subclass would naturally generate an object that is compatible with the type hint due to inheritance and abstraction rules, though in this case I think it's superfluous and confusing to hide the constructor then add a static `create` function, it's reinventing the constructor for no reason.

tomjn replied:
Perhaps if this were a singleton it would make sense as a way to get the global object, but it isn't

dmsnell replied:
> Any subclass would naturally generate an object that is compatible with the type hint due to inheritance and abstraction rules

I’ve really struggled with adding static types here. If we use `self` or `?self` then the only option a subclass has is to return `?WP_Block_Scanner` and not, for instance, `?WP_Block_Processor` (or `?self`).

This is something that gets better with PHP 7.4.

<img width="999" height="529" alt="Screenshot 2025-10-03 at 11 00 20 AM" src="https://github.com/user-attachments/assets/863ce70d-6467-44fb-b650-1590ba88dffd" />

Even with `mixed` as a return type, it still crashes.

> it's superfluous and confusing to hide the constructor then add a static create function, it's reinventing the constructor for no reason.

The difference here is that a static method can return `null` but a constructor can only `throw` if it wants to reject the input arguments. I’m trying to leave room here to explore subclass usage before cutting off the possibility of creating a failable constructor.

At least with HTML it was important to be able to fail on the static creator methods as certain arrangements of arguments cannot be created.

In this case though I think you are right that the scanner class doesn’t need it at all. I’ll remove it. At some point I was hoping to have more harmony between the lower-level and higher-level classes by reusing the static constructor function. People can learn to do it two different ways though.

tomjn replied:
I can foresee an empty string and a string of invalid blocks are the only failure cases, but not sure that this should result in a `null` as that implies an empty post is not possible

dmsnell replied:
this method is now gone and it’s impossible to fail when constructing a block processor.

sirreal commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1244 @@
+	 * @since 6.9.0
+	 *
+	 * @param string $source_text Input document potentially containing block content.
+	 * @return ?self Created block scanner, if successfully created.
```
In the HTML processor we went with `static|null`. This indicates that an instance of the subclass is returned in case of extension:

```suggestion
	 * @return static|null Created block scanner, if successfully created.
```

https://github.com/WordPress/wordpress-develop/blob/2da7a592d643939e0925b821ae4cc2eed29b0db5/src/wp-includes/html-api/class-wp-html-processor.php#L345-L360

---

The `?static` return type annotation does not appear to be supported until PHP 8.0.

dmsnell replied:
hm. I seemed to have confused `self` and `static` once again, as I thought that `static` locked us in to the lexical scope, but `self` was more dynamic for subclassing. confusing. 🤷‍♂️

westonruter replied:
Does a `@return` even make sense here, since constructors don't return values?

dmsnell replied:
this method is now gone.

sirreal commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1244 @@
+	 * @return ?self Created block scanner, if successfully created.
+	 */
+	public static function create( string $source_text ) {
+		return new self( $source_text );
```
Here, I think we want `static` to return a subclass:

```suggestion
		return new static( $source_text );
```

[Demo:](https://3v4l.org/3tTh6)

```php
<?php declare( strict_types = 1 );

class C {
    public static function create_self(): ?self {
        return new self();
    }
    public static function create_static(): ?static {
        return new static();
    }
}

class D extends C {}

var_dump(
    D::create_self(),
    D::create_static(),
);
/* prints:
object(C)#1 (0) {
}
object(D)#2 (0) {
}
*/
```

dmsnell replied:
this method is now gone.

gziolo commented on src/wp-settings.php:
```diff
@@ -268,6 +268,8 @@
 require ABSPATH . WPINC . '/html-api/class-wp-html-processor-state.php';
 require ABSPATH . WPINC . '/html-api/class-wp-html-processor.php';
+require ABSPATH . WPINC . '/class-wp-block-scanner.php';
+require ABSPATH . WPINC . '/html-api/css-helpers.php';
```
I don't think this file is included in the PR. Should it be the block processor file instead?

dmsnell replied:
oops, thanks!

dmsnell replied:
this has been removed

gziolo commented on src/wp-includes/class-wp-block-processor.php:
```diff
@@ -0,0 +1,1380 @@
+	 *
+	 * @param 'visit-html'|'skip-html' $html_spans Optional. Pass `visit-html` to match freeform HTML content
+	 *                                             not surrounded by block delimiters. Defaults to `skip-html`.
+	 * @return bool Whether a block delimiter was matched.
```
What's the difference between a token and a delimiter? The PHPDoc is nearly identical here and for `next_delimiter()`.

dmsnell replied:
I have tried to make the distinctions more obvious in a few places: the class documentation as well as in each method.

What do you think of the changes?

gziolo commented on src/wp-includes/class-wp-block-processor.php:
```diff
@@ -0,0 +1,122 @@
+		return new self( $html );
+	}
+
+	public function next_delimiter( $block_type = null, $html_spans = 'skip-html' ) {
```
The params in this method are different than those in `WP_Block_Scanner`.

dmsnell replied:
this is because I accidentally didn’t remove this file.

dmsnell replied:
this class is gone now, so this is no longer relevant.

gziolo commented:
Can you elaborate on your thinking process for splitting into `WP_Block_Scanner` and `WP_Block_Processor`? I'm not sure whether the latter contains anything unique at this point.

westonruter commented on src/wp-includes/class-wp-block-processor.php:
```diff
@@ -0,0 +1,1380 @@
+	 * @see self::INCOMPLETE_INPUT
+	 * @see self::COMPLETE
+	 *
+	 * @since 6.9.0}
```
```suggestion
	 * @since 6.9.0
```

dmsnell replied:
Removed!

westonruter commented on src/wp-includes/class-wp-block-scanner.php:
```diff
@@ -0,0 +1,1380 @@
+	 *     null                 = $scanner->allocate_and_return_parsed_attributes();
+	 *     JSON_ERROR_CTRL_CHAR = $scanner->get_last_json_error();
+	 *
+	 * @since 6.9.0}
```
```suggestion
	 * @since 6.9.0
```

dmsnell replied:
Removed!

westonruter commented on src/wp-includes/class-wp-block-processor.php:
```diff
@@ -0,0 +1,1380 @@
+	 *
+	 * @see self::state
+	 *
+	 * @since 6.9.0}
```
```suggestion
	 * @since 6.9.0
```

dmsnell replied:
Removed!

dmsnell commented:
> thinking process for splitting into WP_Block_Scanner and WP_Block_Processor?

this is great feedback here because I don’t have a solid feel for splitting or combining them. the core of it is the question of intent: does calling code want to scan the delimiter's or does the calling code want to navigate through the blocks?

this is the same with the HTML API:
 - the Tag Processor only visits real textual tokens, and it visits tokens which may be ignored, ones which may be ignored contextually even, when parsed as HTML.
 - the HTML Processor visits tokens that aren’t in the text itself, implicit openers and closers defined by the parser. it also renames a tag and performs auto-closing.

now with blocks there is far less that the higher-level cousin needs to do:
 - auto-close blocks when parsing errors are present
 - separate `innerHTML` from freeform HTML content
 - create block structure as `parse_blocks()` would produce, which _could_ be built out of the class too

the performance cliff between these two is substantially smaller than with HTML tokenization and parsing. is it worth the separation or not? is the intent-cliff shallow enough? right now I think it’s at least working through the new questions that arise from a single class to see if it seems to fall into place once done, but I’m not ruling out separating them again.

---

separating them out makes it almost trivial to merge `WP_Block_Scanner`, but then it could leave unanswered questions if we _do_ want to merge them because `next_delimiter()` could start behaving differently in a future release.

westonruter commented on src/wp-includes/class-wp-block-processor.php:
```diff
@@ -0,0 +1,122 @@
+<?php
+
+/**
+ * @todo Benchmark extracting blocks against {@link \parse_blocks()}.
```
I presume this will get `@since` and fleshed out phpdoc. It looks like the class is still in a draft state.

dmsnell replied:
yes, and thanks for all this feedback. for the time being I’m making some bigger changes to merge the two interfaces into one. when the design seems solid I will return to ensure that all of the little nitpicky WPCS issues, types, and others are taken care of.

westonruter replied:
Could you change the status of the PR to draft so that reviewers get a better sense of its readiness for review?

dmsnell replied:
done. sorry about that. it _was_ ready for review and then in review became more complicated and scope crept 🙃 and I didn’t think of reverting it

dmsnell replied:
but it’s still in desperate need of valuable high-level feedback, so please do share your thoughts! it’s just not ready for nitpicks, which have their role at a later stage.

dmsnell replied:
this class is now gone as I have merged the two into the `WP_Block_Processor`

westonruter commented on tests/phpunit/tests/block-scanner/wpBlockScanner-BlockProcessing.php:
```diff
@@ -0,0 +1,102 @@
+ * @package WordPress
+ * @subpackage HTML-API
+ *
+ * @since {WP_VERSION}
```
```suggestion
 * @since 6.9.0
```

dmsnell replied:
updated!

dmsnell committed:
Blocks: Introduce WP_Block_Processor for efficiently parsing blocks. (#9105)

The Block Processor follows the HTML API in providing a streaming,
near-zero-overhead, lazy, re-entrant parser for traversing block
structure. This class provides an alternate interface to
`parse_blocks()` which is more amenable to a number of common
server-side operations on posts, such as:

 - Generating an excerpt from only the first N blocks in a post.
 - Determining which block types are present in a post.
 - Determining which posts contain a block of a given type.
 - Generating block supports content for a post.
 - Modifying a single block, or only blocks of a given kind in a post.

Co-authored-by: Sören Wünsch <soean@git.wordpress.org>
Co-authored-by: Tom J Nowell <tjnowell@git.wordpress.org>
Co-authored-by: Weston Ruter <westonruter@git.wordpress.org>
Co-authored-by: Jon Surrell <jonsurrell@git.wordpress.org>
Co-authored-by: Greg Ziółkowski <gziolo@git.wordpress.org>
Github-PR: 9105
Github-PR-URL: https://github.com/WordPress/wordpress-develop/pull/9105
Trac-Ticket: 61401
Trac-Ticket-URL: https://core.trac.wordpress.org/ticket/61401

dmsnell commented:
@gziolo something I was thinking about on my walk this morning was an idea I’ve ignored for a while, but maybe is worth it here: by default we skip freeform HTML content. the way we could signal to visit it is not with a boolean attribute, but using a special block type in `next_delimiter()`.

```php
// visit all explicit block delimiters
$scanner->next_delimiter();

// visit all block delimiters plus freeform blocks.
$scanner->next_delimiter( '*' );

// visit all freeform blocks: a possible special token meaning “no explicit block”
$scanner->next_delimiter( '#freeform' );

// visit all HTML spans, including `innerHTML`
$scanner->next_token();
```

this would collapse the difference between `next_token()` and `next_delimiter()` to whether they visit inner HTML. for any performance ideas about visiting freeform blocks, we could indicate that from within `next_delimiter()` and keep it internal.

I will try and explore this angle.

gziolo commented:
> by default we skip freeform HTML content. the way we could signal to visit it is not with a boolean attribute, but using a special block type in next_delimiter().

That's interesting. It creates a nice logical separation between higher-level `next_delimiter` and lo-level `next_token`.

> ```php
> // visit all freeform blocks: a possible special token meaning “no explicit block”
> $scanner->next_delimiter( '#freeform' );
> ```

We have this concept of freeform content handler in JavaScript which you can even modify with `wp.blocks.setFreeformContentHandlerName` and it defaults to `core/freeform`. I don't think we ever did anything like that on the server, so either `#freeform` or `core/freeform` resonates with me. Maybe folks could even be able to set this alias because I noticed that, for example, in the widgets editor, the freeform handler gets set to `core/html`. It probably is safer to go more meta with `#freeform` or `::freeform` (like [CSS pseudo-elements](https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-elements)).

dmsnell commented:
> We have this concept of freeform content handler in JavaScript which you can even modify with wp.blocks.setFreeformContentHandlerName and it defaults to core/freeform. I don't think we ever did anything like that on the server

That’s something I wish had been more hard-coded and less extensible, based on my work on the two-stage block parsing and loading inside the block editor, both in WordPress and in Tumblr. But I guess JS is different because we have to load _something_ into the editor, whereas on the server a _non-block_ can remain _not a block_.

Since yesterday I’ve been using `*` and deferring an approach which visits only the HTML spans. Something that makes me nervous about using `#freeform` is that we have a relatively short timeline if we want to get this in for 6.9 and properly choosing a name for those seems rushed. It would be an easy thing to add later, I would hope. Something I didn’t mention is why I proposed that syntax:

 - block names cannot start with or have the octothorp character anywhere
 - this mirrors the `#text` and other _token type_ descriptions in the HTML API

When I left off last night I was working on my `extract_block()` method, whose tests were failing. I think there is still some ambiguity in the code around the distinction between `innerHTML` and freeform blocks. What I plan on doing today is to try and do something I’ve been avoiding because of the increased internal state requirements: track HTML spans and pause when we match them. Currently this has all worked by recognizing HTML spans only after matching the next delimiter. With a special state value we can then _visit_ the HTML and move on without re-scanning. There are some complications doing this with incomplete inputs as well, so it’s probably good and time to make this change.

Hopefully that clears up all of the remaining issues.

dmsnell commented:
This has been updated significantly to clarify the interface and to merge the block scanner and block processor into one.

It would be really helpful to have review over the public methods, their docblocks, and the usage in the test suite.

I plan on going back to update the class/module docblock with more substantial general guidance for the class.

Also I plan on doing a separate type annotation pass, so please disregard any comments currently on type annotations.

westonruter commented on src/wp-includes/class-wp-block-processor.php:
```diff
@@ -0,0 +1,1708 @@
+	 * For every open block, in hierarchical order, this stores the byte length
+	 * of the block’s block type in the source text. For HTML spans this is 0.
+	 *
+	 * @seince 6.9.0
```
```suggestion
	 * @since 6.9.0
```

dmsnell replied:
good catch. I hadn’t seen that. gone!

westonruter commented on tests/phpunit/tests/block-processor/wpBlockProcessor.php:
```diff
@@ -0,0 +1,1306 @@
+<?php
+/**
+ * Unit tests covering WP_Block_Scanner functionality.
```
```suggestion
 * Unit tests covering WP_Block_Processor functionality.
```

dmsnell replied:
Updated!

westonruter commented on src/wp-includes/class-wp-block-processor.php:
```diff
@@ -0,0 +1,1708 @@
+	 *
+	 * @since 6.9.0
+	 *
+	 * @return array
```
```suggestion
	 * @return string[]
```

dmsnell replied:
updated!

westonruter reviewed (commented):
Just a couple nits

dmsnell commented:
With significant haste in the eleventh hour I have finished making some major changes to this proposal:

 - There is now only a single `WP_Block_Processor` class (thanks @swissspidy for renaming the PR).
 - This singular class differentiates between freeform content and inner HTML, making it possible for calling code to inquire which is which.
   - A new block type wildcard `*` can be passed to the `next_` functions to ensure that they visit top-level freeform content.
 - The docs have all been updated, names have been changed, and type annotations have been added.
 - The processor treats HTML spans like void blocks; previously it created a separate opener and closer.
 - Almost all references to `static::` have been replaced by `self::`. After consideration, as a default first step I think it would be wise to treat most of a this class as `final` and discourage subclasses from reimplementing methods like `next_token()` the way subclasses do in the HTML API. Without _needing_ a higher-level subclass this makes it easier.
   - The exception are a few methods intended to be replaced by subclasses, including the missing lazy JSON parser for `static::get_attributes()`.

Personally I’m quite happy with this revision and I thank you all for your input on this. I think it’s in a much more coherent and usable state now. At this point I believe it’s ready for final review and merge, though I think it would be best only to block merging now for design issues with the class, the methods, the things which must be maintained _ad infinitum_. For other issues, I would prefer if we can address them during the beta phase after merging. 

Of course, I invite all feedback and will address things if I can in a timely manner, but right now I would love to see this out there in `trunk` to get it in use and testing.
